tests/tests/test_filetransfer.py

```diff
@@ -281,4 +323,406 @@ def test_filetransfer(self, enterprise_no_client):
 
         authtoken = auth.get_auth_token()
 
-        super().test_filetransfer(devices[0], authtoken)
+        super().test_filetransfer(devices[0], authtoken, content_assertion="ServerURL")
+
+
+class TestFileTransferLimits:
+    def test_filetransfer_limits_setup(self, standard_setup_one_client):
+        """Tests the file transfer features"""
+        # accept the device
+        devauth.accept_devices(1)
+
+        # list of devices
+        devices = list(
+            set([device["id"] for device in devauth.get_devices_status("accepted")])
+        )
+        assert 1 == len(devices)
+
+        # wait for the device to connect via websocket
+        auth = authentication.Authentication()
+        wait_for_connect(auth, devices[0])
+
+        # device ID and auth token
+        devid = devices[0]
+        authtoken = auth.get_auth_token()
+
+        mender_device = MenderDevice(standard_setup_one_client.get_mender_clients()[0])
+        mender_device.ssh_is_opened()
+        return mender_device, devid, authtoken
+
+    def test_filetransfer_limits_upload(self, standard_setup_one_client):
+        mender_device, devid, authtoken = self.test_filetransfer_limits_setup(
+            standard_setup_one_client
+        )
+        """Tests the file transfer features with limits"""
+        set_limits(
+            mender_device,
+            {
+                "Enabled": True,
+                "FileTransfer": {"Chroot": "/var/lib/mender/filetransfer",},
+            },
+        )
+        mender_device.run("mkdir -p /var/lib/mender/filetransfer")
+        time.sleep(4)  # wait for mender-connect re-connect
+
+        logger.info(
+            "-- testcase: File Transfer limits: file outside chroot; upload forbidden"
+        )
+        f = NamedTemporaryFile(delete=False)
+        for i in range(40 * 1024):
+            f.write(os.urandom(1024))
+        f.close()
+        r = upload_file("/usr/random.bin", open(f.name, "rb"), devid, authtoken,)
+
+        assert r.status_code == 400, r.json()
+        assert (
+            r.json().get("error")
+            == "access denied: the target file path is outside chroot"
+        )
+
+        set_limits(
+            mender_device,
+            {
+                "Enabled": True,
+                "FileTransfer": {
+                    "Chroot": "/var/lib/mender/filetransfer",
+                    "FollowSymLinks": True,  # in the image /var/lib/mender is a symlink
+                },
+            },
+        )
+        time.sleep(4)  # wait for mender-connect re-connect
```

Comment:
I would replace it with ` wait_for_connect(...)`
It applies to all the sleep(4)

Reply:
I added wait_for_connect to set_limits, removed the sleeps, and changed set_limits to accept auth and devid as parameters, that also meant changing the return from setup_test_filetransfer_limits.